// readmeio/dist/src/lib/metrics-log.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.metricsAPICall = void 0;
var node_fetch_1 = __importDefault(require("node-fetch"));
var timeout_signal_1 = __importDefault(require("timeout-signal"));
var package_json_1 = __importDefault(require("../../package.json"));
var config_1 = __importDefault(require("../config"));
function getLogIds(body) {
    if (Array.isArray(body)) {
        return body.map(function (value) { return value._id; });
    }
    return body._id;
}
function metricsAPICall(readmeAPIKey, body, fireAndForget) {
    if (fireAndForget === void 0) { fireAndForget = false; }
    var signal = (0, timeout_signal_1["default"])(config_1["default"].timeout);
    var encodedKey = Buffer.from("".concat(readmeAPIKey, ":")).toString('base64');
    var makeRequest = function () {
        return (0, node_fetch_1["default"])(new URL('/v1/request', config_1["default"].host), {
            method: 'post',
            body: JSON.stringify(body),
            headers: {
                Authorization: "Basic ".concat(encodedKey),
                'Content-Type': 'application/json',
                'User-Agent': "".concat(package_json_1["default"].name, "/").concat(package_json_1["default"].version)
            },
            signal: signal
        })["finally"](function () {
            timeout_signal_1["default"].clear(signal);
        });
    };
    if (fireAndForget) {
        makeRequest();
        return Promise.resolve({
            ids: getLogIds(body)
        });
    }
    return makeRequest().then(function (response) {
        return {
            response: response,
            ids: getLogIds(body)
        };
    });
}
exports.metricsAPICall = metricsAPICall;
